docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md

```diff
@@ -1,139 +1,243 @@
-## This document lists known breaking changes in Roslyn after .NET 6 all the way to .NET 7.
+# This document lists known breaking changes in Roslyn after .NET 6 all the way to .NET 7.
 
-1. <a name="1"></a>In Visual Studio 17.1, the contextual keyword `var` cannot be used as an explicit lambda return type.
+## Foreach enumerator as a ref struct
 
-    ```csharp
-    using System;
+***Introduced in .NET SDK 6.0.300, Visual Studio 2022 version 17.2.***. A `foreach` using a ref struct enumerator type reports an error if the language version is set to 7.3 or earlier.
 
-    F(var () => default);  // error: 'var' cannot be used as an explicit lambda return type
-    F(@var () => default); // ok
+This fixes a bug where the feature was supported in newer compilers targeting a version of C# prior to its support.
 
-    static void F(Func<var> f) { }
+Possible workarounds are:
 
-    class var { }
-    ```
+1. Change the `ref struct` type to a `struct` or `class` type.
+1. Upgrade the `<LangVersion>` element to 7.3 or later.
 
-2. <a name="2"></a>In Visual Studio 17.1, indexers that take an interpolated string handler and require the receiver as an input for the constructor cannot be used in an object initializer.
+## Async `foreach` prefers pattern based `DisposeAsync` to and explicit interface implementation of `IAsyncDisposable.DisposeAsync()`
 
-    ```cs
-    using System.Runtime.CompilerServices;
-    _ = new C { [$""] = 1 }; // error: Interpolated string handler conversions that reference the instance being indexed cannot be used in indexer member initializers.
+***Introduced in .NET SDK 6.0.300, Visual Studio 2022 version 17.2.***. An async `foreach` prefers to bind using a pattern-based `DisposeAsync()` method rather than `IAsyncDisposable.DisposeAsync()`.
 
-    class C
+For instance, the `DisposeAsync()` will be picked, rather than the `IAsyncEnumerator<int>.DisposeAsync()` method on `AsyncEnumerator`:
+
+```csharp
+await foreach (var i in new AsyncEnumerable())
+{
+}
+
+struct AsyncEnumerable
+{
+    public AsyncEnumerator GetAsyncEnumerator() => new AsyncEnumerator();
+}
+
+struct AsyncEnumerator : IAsyncDisposable
+{
+    public int Current => 0;
+    public async ValueTask<bool> MoveNextAsync()
     {
-        public int this[[InterpolatedStringHandlerArgument("")] CustomHandler c]
-        {
-            get => throw null;
-            set => throw null;
-        }
+        await Task.Yield();
+        return false;
     }
+    public async ValueTask DisposeAsync()
+    {
+        Console.WriteLine("PICKED");
+        await Task.Yield();
+    }
+    ValueTask IAsyncDisposable.DisposeAsync() => throw null; // no longer picked
+}
+```
+
+This change fixes a spec violation where the public `DisposeAsync` method is visible on the declared type, whereas the explicit interface implementation is only visible using a reference to the interface type.
+
+To workaround this error, remove the pattern based `DisposeAsync` method from your type.
```

Comment:
This change isn't yet available in the preview I'm running. I'll verify once the next internal preview is available, or you can verify.